bpo-28879 : add date if missing in smtplib.send_message

elafontaine opened the pull request:
Hi I've tried to change the patch submitted on bpo-28879 to be on a github PR.

It's my first time contributing through PR, please let me know what I did wrong.

Regards,
Eric Lafontaine

<!-- issue-number: bpo-28879 -->
https://bugs.python.org/issue28879
<!-- /issue-number -->

elafontaine committed:
bpo-28879 : add date if missing in smtplib.send_message

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

elafontaine commented:
oh wow... this is really a nice setup you've got running there!  I see that some check failed for some trailing whitespace in my files... 

Unfortunatly, I seems to be unable to look at the log anymore :S.

Anyway, how could I see it again?

elafontaine commented:
Ok got to find how to see it again.  Now to understand what "trailing whitespace" is supposed to be for python and for... C (but I didn't touch the C part...)

Mariatta commented:
The clue is at this line in travis log: (https://travis-ci.org/python/cpython/jobs/252242241#L494)

```
library/smtplib.rst:499: trailing whitespace
```

There is an extra space after the word `no` in line 499 of `smtplib.rst`.
The space should be removed.

mlouielu commented on Lib/test/test_smtplib.py:
```diff
@@ -549,6 +549,36 @@ def testSendMessageMultipleResentRaises(self):
+        self.assertRegex(debugout, Date)
+
+
+
```
This only need two blankline to spacing class and class

mlouielu commented on Lib/smtplib.py:
```diff
@@ -928,6 +928,12 @@ def send_message(self, msg, from_addr=None, to_addrs=None,
+        if msg.get('Date',None) is None:
+            msg['Date'] = email.utils.formatdate()
+        
+        
```
This only need one blankline.

mlouielu reviewed (changes requested):
Hello, welcome to CPython community, thanks for contributing with CPython, and here is some coding style issue with PEP8 that can be fixed in your PR!

elafontaine committed:
bpo-28879 : add date if missing in smtplib.send_message part 2

Fix the whitespace and spacing

elafontaine commented:
I've followed the recommandation (thanks @Mariatta @mlouielu ).  How can I re-trigger the CI now that the commit seems to be in the PR?

mlouielu commented:
@lafe3402 CI will automatically re-trigger when you push your commit to GitHub. 

Also, I would recommend you to read this article: [How to write a git commit message](https://chris.beams.io/posts/git-commit/). This article is very useful that it tells how to write a good commit message. For example, your second commit message would be much better to changed to `Addressed somebody's comments` or `Fix coding style blankline`.

elafontaine commented:
yeah... I though that I would have to make another PR request... so I've put what I've done in the subsection of the commit (it's in the expand the commit message if you expand it...).

Thanks a lot for the feedback :) you're really pro-active on the comment and I appreciate that.

Anything else I could do to ease the process?

Regards,
Eric Lafontaine

mlouielu commented on Lib/smtplib.py:
```diff
@@ -928,6 +928,11 @@ def send_message(self, msg, from_addr=None, to_addrs=None,
             raise ValueError("message has more than one 'Resent-' header block")
+
+        # RFC 5322 section 3.6, 4th Paragraph
+        if msg.get('Date',None) is None:
```
Here missing a whitespace after the `,`.

mlouielu commented on Lib/test/test_smtplib.py:
```diff
@@ -549,6 +549,34 @@ def testSendMessageMultipleResentRaises(self):
+        m['CC'] = 'Sally, Fred'
+        m['Bcc'] = 'John Root <root@localhost>, "Dinsdale" <warped@silly.walks.com>'
+        smtp = smtplib.SMTP(HOST, self.port, local_hostname='localhost', timeout=3)
+        smtp.send_message(m)
```
Maybe we can mock up the `email.utils.formatdate()` to ensure we have the same result to tests.

mlouielu commented on Lib/test/test_smtplib.py:
```diff
@@ -549,6 +549,34 @@ def testSendMessageMultipleResentRaises(self):
+        mexpect = '%s%s\n%s' % (MSG_BEGIN, m.as_string(), MSG_END)
+        self.assertEqual(self.output.getvalue(), mexpect)
+        debugout = smtpd.DEBUGSTREAM.getvalue()
+        current_moment = email.utils.formatdate()
```
see above to mock this up.

elafontaine replied:
Thanks for the review :) I'll work on it tonight, my wife won't be home so I'll have nothing better to do :).

I'm not sure what you mean though on the "mock this up" but I guess you mean to use @mock(email.utils.formatdate) ?

I'll have to read to remember how to do it correctly if it's what you meant.

elafontaine committed:
bpo-28879 : Mock the date on the Test Case + Fix Typo

Following the discussion on the bpo-28879, I've put space between commas and the variable to be compliant and mocked the "Date" call being made to the send_message function

elafontaine committed:
bpo-28879 : Fix test case with less '\'

elafontaine commented:
I'm sorry about this last re-commit, I can't run the test right now so I'm using the CI :$.

mlouielu commented:
@lafe3402 I can't pass the unittest on my laptop, too. You will need to check about the failed problem.

elafontaine committed:
bpo-28879 : escape the date example string

elafontaine commented:
@mlouielu , thanks for your patience, I had trouble finding some free time.

elafontaine commented:
Hi @mlouielu,

Would you have some feedback?  Just curious to see if there's something better I could've had done.

Regards,

elafontaine commented:
@mlouielu @bitdancer,

Could you let me know what to do to accelerate the closure of this ticket?

bitdancer commented:
Find me some free time somewhere?  I'm sorry, I just don't have the cycles to look at this, nor will I any time soon :( :(

elafontaine commented:
@bitdancer, Anything I could do to help you?  

I've been re-reading my question and I just don't want it to be misinterpreted as a "I want my request to get in!" kind of request.  I'm genuinely trying to help you guys.  I love python and I would love to contribute more.  

Also, David, are you the only one who can approve?  If yes, should I learn email modules and try to alleviate some of the pain?  

Again, sorry for asking here, but I would like more meat on how to contribute or how to learn what I need to contribute XD.

Regards,
Eric Lafontaine

Commit:
Adding news file to advertise changes done by this PR.
